src/compiler/moduleSpecifiers.ts

```diff
@@ -832,17 +847,54 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode = MatchingMode.Exact): { moduleFileToTry: string; } | undefined {
+function getOutputPath(targetFilePath: string, options: CompilerOptions, commonDir: string) {
+    if (options.emitDeclarationOnly) {
+        return;
+    }
+    return changeExtension(
+        options.outDir ?
+            resolvePath(
+                options.outDir,
+                getRelativePathFromDirectory(commonDir, targetFilePath, /*ignoreCase*/ false),
+            ) :
+            targetFilePath,
+        getOutputExtension(targetFilePath, options),
+    );
+}
+
+function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, host: ModuleSpecifierResolutionHost, getCanonicalFileName: GetCanonicalFileName, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode, isImports: boolean): { moduleFileToTry: string; } | undefined {

```

Comment:
I introduced `isImports` parameter here since I planned to use it - but should I? Should this behavior be only allowed for `package.json#imports`? Or should self-referencing based on `package.json#exports` also be allowed?